[Bug]: Parameters in element are passed by reference

### Description

Parameters in `zyn::element` components are passed by reference to the real type of the parameter.

### Steps to Reproduce

Use the following code :
```rust
#[zyn::element]
fn crate_prefix(in_crate: bool) -> zyn::TokenStream {
    if in_crate {
        zyn! {crate}
    } else {
        zyn! {::functional_type}
    }
}
```

### Expected Behavior

`in_crate` should be a boolean

### Actual Behavior

`in_crate` is passed as a `&bool` :
```rust
error[E0308]: mismatched types
 --> functional-type-macros/src/lib.rs:9:8
  |
9 |     if in_crate {
  |        ^^^^^^^^ expected `bool`, found `&bool`
  |
help: consider dereferencing the borrow
  |
9 |     if *in_crate {
  |        +
```

### Zyn Version

0.5.4

### Rust Version

rustc 1.93.1 (01f6ddf75 2026-02-11)

### Operating System

Linux

### Additional Context

_No response_

--- a/crates/zyn-core/src/extract/attr.rs
+++ b/crates/zyn-core/src/extract/attr.rs
@@ -15,16 +15,16 @@ use super::FromInput;
 ///     // cfg.my_field — accessed via Deref
 /// }
 /// ```
-pub struct Attr<T: FromInput>(T);
+pub struct Attr<T>(T);
 
-impl<T: FromInput> Attr<T> {
+impl<T> Attr<T> {
     /// Consumes the wrapper and returns the inner value.
     pub fn inner(self) -> T {
         self.0
     }
 }
 
-impl<T: FromInput> std::ops::Deref for Attr<T> {
+impl<T> std::ops::Deref for Attr<T> {
     type Target = T;
 
     fn deref(&self) -> &Self::Target {
@@ -32,14 +32,17 @@ impl<T: FromInput> std::ops::Deref for Attr<T> {
     }
 }
 
-impl<T: FromInput> std::ops::DerefMut for Attr<T> {
+impl<T> std::ops::DerefMut for Attr<T> {
     fn deref_mut(&mut self) -> &mut Self::Target {
         &mut self.0
     }
 }
 
-impl<T: FromInput> FromInput for Attr<T> {
-    fn from_input(input: &Input) -> crate::Result<Self> {
+impl<'i, T> FromInput<'i> for Attr<T>
+where
+    T: FromInput<'i>,
+{
+    fn from_input(input: &'i Input) -> crate::Result<Self> {
         T::from_input(input).map(Attr)
     }
 }

--- a/crates/zyn-core/src/extract/data.rs
+++ b/crates/zyn-core/src/extract/data.rs
@@ -96,8 +96,8 @@ impl<T: FromData> std::ops::DerefMut for Data<T> {
     }
 }
 
-impl<T: FromData> FromInput for Data<T> {
-    fn from_input(input: &Input) -> crate::Result<Self> {
+impl<'i, T: FromData> FromInput<'i> for Data<T> {
+    fn from_input(input: &'i Input) -> crate::Result<Self> {
         match input {
             Input::Derive(d) => T::from_data(d.data.clone()).map(Data),
             _ => Err(mark::error("Data extractor requires derive input")

--- a/crates/zyn-core/src/extract/fields.rs
+++ b/crates/zyn-core/src/extract/fields.rs
@@ -76,8 +76,8 @@ impl<T: FromFields> std::ops::DerefMut for Fields<T> {
     }
 }
 
-impl<T: FromFields> FromInput for Fields<T> {
-    fn from_input(input: &Input) -> crate::Result<Self> {
+impl<'i, T: FromFields> FromInput<'i> for Fields<T> {
+    fn from_input(input: &'i Input) -> crate::Result<Self> {
         let raw = match input {
             Input::Derive(d) => match &d.data {
                 syn::Data::Struct(s) => s.fields.clone(),

--- a/crates/zyn-core/src/extract/mod.rs
+++ b/crates/zyn-core/src/extract/mod.rs
@@ -42,7 +42,10 @@
 //! struct MyExtractor(String);
 //!
 //! impl FromInput for MyExtractor {
-//!     fn from_input(input: &Input) -> Result<Self> {
+//!     fn from_input<'i>(input: &'i Input) -> Result<Self>
+//!     where
+//!         Self: 'i,
+//!     {
 //!         Ok(MyExtractor(input.ident().to_string()))
 //!     }
 //! }
@@ -72,24 +75,29 @@ pub use variants::*;
 /// (derive or item). Built-in impls exist for `Ident`, `Generics`, and
 /// `Visibility`. The `#[element]` macro uses this trait to auto-resolve
 /// extractor parameters.
-pub trait FromInput: Sized {
-    fn from_input(input: &Input) -> crate::Result<Self>;
+///
+/// The lifetime `'i` ties `Self` to the input reference, enabling both
+/// owned impls (`impl<'i> FromInput<'i> for syn::Ident`) and borrowed impls
+/// (`impl<'i> FromInput<'i> for &'i syn::Ident`). Owned types satisfy any
+/// `'i` because they are `'static`.
+pub trait FromInput<'i>: Sized {
+    fn from_input(input: &'i Input) -> crate::Result<Self>;
 }
 
 /// Generic extractor wrapper — delegates to `T::from_input`.
 ///
 /// Use this in element parameters to extract any `FromInput` type
 /// without giving it a more specific semantic role like `Attr` or `Fields`.
-pub struct Extract<T: FromInput>(T);
+pub struct Extract<T>(T);
 
-impl<T: FromInput> Extract<T> {
+impl<T> Extract<T> {
     /// Consumes the wrapper and returns the inner value.
     pub fn inner(self) -> T {
         self.0
     }
 }
 
-impl<T: FromInput> std::ops::Deref for Extract<T> {
+impl<T> std::ops::Deref for Extract<T> {
     type Target = T;
 
     fn deref(&self) -> &Self::Target {
@@ -97,33 +105,54 @@ impl<T: FromInput> std::ops::Deref for Extract<T> {
     }
 }
 
-impl<T: FromInput> std::ops::DerefMut for Extract<T> {
+impl<T> std::ops::DerefMut for Extract<T> {
     fn deref_mut(&mut self) -> &mut Self::Target {
         &mut self.0
     }
 }
 
-impl<T: FromInput> FromInput for Extract<T> {
-    fn from_input(input: &Input) -> crate::Result<Self> {
+impl<'i, T> FromInput<'i> for Extract<T>
+where
+    T: FromInput<'i>,
+{
+    fn from_input(input: &'i Input) -> crate::Result<Self> {
         T::from_input(input).map(Extract)
     }
 }
 
-impl FromInput for proc_macro2::Ident {
-    fn from_input(input: &Input) -> crate::Result<Self> {
+impl<'i> FromInput<'i> for proc_macro2::Ident {
+    fn from_input(input: &'i Input) -> crate::Result<Self> {
         Ok(input.ident().clone())
     }
 }
 
-impl FromInput for syn::Generics {
-    fn from_input(input: &Input) -> crate::Result<Self> {
+impl<'i> FromInput<'i> for &'i proc_macro2::Ident {
+    fn from_input(input: &'i Input) -> crate::Result<Self> {
+        Ok(input.ident())
+    }
+}
+
+impl<'i> FromInput<'i> for syn::Generics {
+    fn from_input(input: &'i Input) -> crate::Result<Self> {
         Ok(input.generics().clone())
     }
 }
 
-impl FromInput for syn::Visibility {
-    fn from_input(input: &Input) -> crate::Result<Self> {
+impl<'i> FromInput<'i> for &'i syn::Generics {
+    fn from_input(input: &'i Input) -> crate::Result<Self> {
+        Ok(input.generics())
+    }
+}
+
+impl<'i> FromInput<'i> for syn::Visibility {
+    fn from_input(input: &'i Input) -> crate::Result<Self> {
         Ok(input.vis().clone())
+    }
+}
+
+impl<'i> FromInput<'i> for &'i syn::Visibility {
+    fn from_input(input: &'i Input) -> crate::Result<Self> {
+        Ok(input.vis())
     }
 }
 

--- a/crates/zyn-core/src/extract/variants.rs
+++ b/crates/zyn-core/src/extract/variants.rs
@@ -40,8 +40,8 @@ impl std::ops::DerefMut for Variants {
     }
 }
 
-impl FromInput for Variants {
-    fn from_input(input: &Input) -> crate::Result<Self> {
+impl<'i> FromInput<'i> for Variants {
+    fn from_input(input: &'i Input) -> crate::Result<Self> {
         match input {
             Input::Derive(d) => match &d.data {
                 syn::Data::Enum(e) => Ok(Variants(e.variants.clone())),

--- a/crates/zyn-core/src/types/derive.rs
+++ b/crates/zyn-core/src/types/derive.rs
@@ -2,8 +2,8 @@ use super::Input;
 use crate::extract::FromInput;
 use crate::mark;
 
-impl FromInput for syn::DeriveInput {
-    fn from_input(input: &Input) -> crate::Result<Self> {
+impl<'i> FromInput<'i> for syn::DeriveInput {
+    fn from_input(input: &'i Input) -> crate::Result<Self> {
         match input {
             Input::Derive(d) => Ok(d.clone()),
             _ => Err(mark::error("expected derive input")
@@ -13,8 +13,8 @@ impl FromInput for syn::DeriveInput {
     }
 }
 
-impl FromInput for syn::DataStruct {
-    fn from_input(input: &Input) -> crate::Result<Self> {
+impl<'i> FromInput<'i> for syn::DataStruct {
+    fn from_input(input: &'i Input) -> crate::Result<Self> {
         match input {
             Input::Derive(d) => match &d.data {
                 syn::Data::Struct(s) => Ok(s.clone()),
@@ -27,8 +27,8 @@ impl FromInput for syn::DataStruct {
     }
 }
 
-impl FromInput for syn::DataEnum {
-    fn from_input(input: &Input) -> crate::Result<Self> {
+impl<'i> FromInput<'i> for syn::DataEnum {
+    fn from_input(input: &'i Input) -> crate::Result<Self> {
         match input {
             Input::Derive(d) => match &d.data {
                 syn::Data::Enum(e) => Ok(e.clone()),
@@ -41,8 +41,8 @@ impl FromInput for syn::DataEnum {
     }
 }
 
-impl FromInput for syn::DataUnion {
-    fn from_input(input: &Input) -> crate::Result<Self> {
+impl<'i> FromInput<'i> for syn::DataUnion {
+    fn from_input(input: &'i Input) -> crate::Result<Self> {
         match input {
             Input::Derive(d) => match &d.data {
                 syn::Data::Union(u) => Ok(u.clone()),

--- a/crates/zyn-core/src/types/item.rs
+++ b/crates/zyn-core/src/types/item.rs
@@ -2,8 +2,8 @@ use super::Input;
 use crate::extract::FromInput;
 use crate::mark;
 
-impl FromInput for syn::Item {
-    fn from_input(input: &Input) -> crate::Result<Self> {
+impl<'i> FromInput<'i> for syn::Item {
+    fn from_input(input: &'i Input) -> crate::Result<Self> {
         match input {
             Input::Item(v) => Ok(v.clone()),
             _ => Err(mark::error("expected item input")
@@ -15,8 +15,8 @@ impl FromInput for syn::Item {
 
 macro_rules! impl_from_input_item {
     ($ty:ty, $variant:ident, $msg:literal) => {
-        impl FromInput for $ty {
-            fn from_input(input: &Input) -> crate::Result<Self> {
+        impl<'i> FromInput<'i> for $ty {
+            fn from_input(input: &'i Input) -> crate::Result<Self> {
                 match input {
                     Input::Item(syn::Item::$variant(v)) => Ok(v.clone()),
                     _ => Err(mark::error($msg).span(input.span()).build()),
@@ -45,8 +45,8 @@ impl_from_input_item!(syn::ItemTrait, Trait, "expected trait item input");
 impl_from_input_item!(syn::ItemType, Type, "expected type item input");
 impl_from_input_item!(syn::ItemUse, Use, "expected use item input");
 
-impl FromInput for syn::ItemStruct {
-    fn from_input(input: &Input) -> crate::Result<Self> {
+impl<'i> FromInput<'i> for syn::ItemStruct {
+    fn from_input(input: &'i Input) -> crate::Result<Self> {
         match input {
             Input::Item(syn::Item::Struct(v)) => Ok(v.clone()),
             Input::Derive(d) => match &d.data {
@@ -70,8 +70,8 @@ impl FromInput for syn::ItemStruct {
     }
 }
 
-impl FromInput for syn::ItemEnum {
-    fn from_input(input: &Input) -> crate::Result<Self> {
+impl<'i> FromInput<'i> for syn::ItemEnum {
+    fn from_input(input: &'i Input) -> crate::Result<Self> {
         match input {
             Input::Item(syn::Item::Enum(v)) => Ok(v.clone()),
             Input::Derive(d) => match &d.data {
@@ -95,8 +95,8 @@ impl FromInput for syn::ItemEnum {
     }
 }
 
-impl FromInput for syn::ItemUnion {
-    fn from_input(input: &Input) -> crate::Result<Self> {
+impl<'i> FromInput<'i> for syn::ItemUnion {
+    fn from_input(input: &'i Input) -> crate::Result<Self> {
         match input {
             Input::Item(syn::Item::Union(v)) => Ok(v.clone()),
             Input::Derive(d) => match &d.data {

--- a/crates/zyn-derive/src/attribute/emit.rs
+++ b/crates/zyn-derive/src/attribute/emit.rs
@@ -1,3 +1,4 @@
+use zyn_core::proc_macro2;
 use zyn_core::proc_macro2::TokenStream;
 use zyn_core::quote::quote;
 use zyn_core::syn;
@@ -223,11 +224,22 @@ pub fn from_arg(
 pub fn from_input(
     name: &syn::Ident,
     meta: &StructMeta,
-    impl_generics: &syn::ImplGenerics<'_>,
-    ty_generics: &syn::TypeGenerics<'_>,
+    generics: &syn::Generics,
     where_clause: Option<&syn::WhereClause>,
 ) -> TokenStream {
     let attr_name = meta.attr_name.as_deref().unwrap();
+    let (_, ty_generics, _) = generics.split_for_impl();
+    let mut extended = generics.clone();
+
+    extended.params.insert(
+        0,
+        syn::GenericParam::Lifetime(syn::LifetimeParam::new(syn::Lifetime::new(
+            "'__i",
+            proc_macro2::Span::call_site(),
+        ))),
+    );
+
+    let (impl_generics, _, _) = extended.split_for_impl();
 
     let unique_check = if meta.unique {
         let msg = format!("only one #[{attr_name}(...)] allowed");
@@ -264,8 +276,8 @@ pub fn from_input(
     };
 
     quote! {
-        impl #impl_generics ::zyn::FromInput for #name #ty_generics #where_clause {
-            fn from_input(input: &::zyn::Input) -> ::zyn::Result<Self> {
+        impl #impl_generics ::zyn::FromInput<'__i> for #name #ty_generics #where_clause {
+            fn from_input(input: &'__i ::zyn::Input) -> ::zyn::Result<Self> {
                 let matches: ::std::vec::Vec<_> = input.attrs().iter()
                     .filter(|a| a.path().is_ident(#attr_name))
                     .collect();
@@ -286,7 +298,6 @@ pub fn about(
     where_clause: Option<&syn::WhereClause>,
 ) -> TokenStream {
     let attr_name = meta.attr_name.as_deref().unwrap();
-
     let header = match &meta.about {
         Some(about) => format!("#[{attr_name}(...)]: {about}"),
         None => format!("#[{attr_name}(...)]"),

--- a/crates/zyn-derive/src/attribute/structs/mod.rs
+++ b/crates/zyn-derive/src/attribute/structs/mod.rs
@@ -32,8 +32,7 @@ pub fn expand(input: DeriveInput) -> TokenStream {
         Some(emit::from_input(
             name,
             &struct_meta,
-            &impl_generics,
-            &ty_generics,
+            &input.generics,
             where_clause,
         ))
     } else {

--- a/crates/zyn-derive/src/common/extractors.rs
+++ b/crates/zyn-derive/src/common/extractors.rs
@@ -12,7 +12,7 @@ pub fn bindings(
         .zip(types.iter())
         .map(|(name, ty)| {
             quote! {
-                let #name = match <#ty as ::zyn::FromInput>::from_input(#input_expr) {
+                let #name = match <#ty as ::zyn::FromInput<'_>>::from_input(#input_expr) {
                     ::std::result::Result::Ok(v) => v,
                     ::std::result::Result::Err(e) => {
                         diagnostics = diagnostics.add(e);

--- a/crates/zyn-derive/src/macros/derive.rs
+++ b/crates/zyn-derive/src/macros/derive.rs
@@ -150,7 +150,7 @@ fn expand_derive(item: ItemFn, args: DeriveArgs) -> TokenStream {
     }
 
     let diagnostic_macros = crate::common::diagnostics::macros();
-    let input_expr = quote! { &::zyn::Input::from(input.clone()) };
+    let input_expr = quote! { &input };
     let extractor_bindings =
         crate::common::extractors::bindings(&extractor_names, &extractor_types, &input_expr);
 
@@ -159,7 +159,7 @@ fn expand_derive(item: ItemFn, args: DeriveArgs) -> TokenStream {
         #derive_attr
         #[allow(unreachable_code)]
         pub fn #fn_name(__zyn_input: proc_macro::TokenStream) -> proc_macro::TokenStream {
-            let input = ::zyn::parse_input!(__zyn_input as ::zyn::syn::DeriveInput);
+            let input: ::zyn::Input = ::zyn::parse_input!(__zyn_input as ::zyn::syn::DeriveInput).into();
 
             let __zyn_result: ::zyn::Output = (|| {
                 let mut diagnostics = ::zyn::mark::new();
@@ -169,7 +169,6 @@ fn expand_derive(item: ItemFn, args: DeriveArgs) -> TokenStream {
                 #(#extractor_bindings)*
 
                 let __body = #body;
-
                 ::zyn::Output::new()
                     .tokens(__body)
                     .diagnostic(diagnostics)

--- a/tests/element/extract.rs
+++ b/tests/element/extract.rs
@@ -1,5 +1,66 @@
 use zyn::quote::quote;
 
+// ── Reference parameter tests ─────────────────────────────────────────────────
+
+#[zyn::element]
+fn ref_ident_element(#[zyn(input)] ident: &zyn::syn::Ident) -> zyn::TokenStream {
+    zyn::zyn!(
+        const NAME: &str = { { ident | str } };
+    )
+}
+
+#[test]
+fn ref_ident() {
+    let input: zyn::Input = zyn::parse!("struct MyStruct;").unwrap();
+    let output = zyn::zyn!(@ref_ident_element());
+    let expected = quote!(
+        const NAME: &str = "MyStruct";
+    );
+    zyn::assert_tokens!(output, expected);
+}
+
+#[zyn::element]
+fn ref_generics_element(#[zyn(input)] generics: &zyn::syn::Generics) -> zyn::TokenStream {
+    let count = generics.params.len();
+    zyn::zyn!(
+        const COUNT: usize = { { count } };
+    )
+}
+
+#[test]
+fn ref_generics() {
+    let input: zyn::Input = zyn::parse!("struct Foo<A, B, C>;").unwrap();
+    let output = zyn::zyn!(@ref_generics_element());
+    let expected = quote!(
+        const COUNT: usize = 3usize;
+    );
+    zyn::assert_tokens!(output, expected);
+}
+
+#[zyn::element]
+fn ref_ident_and_generics_element(
+    #[zyn(input)] ident: &zyn::syn::Ident,
+    #[zyn(input)] generics: &zyn::syn::Generics,
+) -> zyn::TokenStream {
+    let name_str = ident.to_string();
+    let param_count = generics.params.len();
+    zyn::zyn!(
+        const NAME: &str = { { name_str | str } };
+        const PARAMS: usize = { { param_count } };
+    )
+}
+
+#[test]
+fn ref_ident_and_generics() {
+    let input: zyn::Input = zyn::parse!("struct Point<X, Y>;").unwrap();
+    let output = zyn::zyn!(@ref_ident_and_generics_element());
+    let expected = quote!(
+        const NAME: &str = "Point";
+        const PARAMS: usize = 2usize;
+    );
+    zyn::assert_tokens!(output, expected);
+}
+
 #[derive(zyn::Attribute)]
 #[zyn("my_attr")]
 struct TestAttr {

